sdk/typespec/typespec_client_core/src/macros.rs

```diff
@@ -1,6 +1,100 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+/// The following macro invocation:
+/// ```
+/// # #[macro_use] extern crate typespec_client_core;
+/// create_enum!(Words, (Chicken, "Chicken"), (White, "White"), (Yellow, "Yellow"));
+/// ```
+/// Turns into a struct where each variant can be turned into and construct from the corresponding string.
+#[macro_export]
+macro_rules! create_enum {
+    ($(#[$meta:meta])* $name:ident, $(($variant:ident, $value:expr)), *) => (
+        $(#[$meta])*
+        #[derive(Debug, PartialEq, Eq, PartialOrd, Clone, Copy)]
+        pub enum $name {
+            $(
+                $variant,
+            )*
+        }
+
+        impl ::std::convert::From<$name> for &'static str {
+            fn from(e: $name) -> Self {
+                match e {
+                    $(
+                        $name::$variant => $value,
+                    )*
+                }
+            }
+        }
+
+        impl $crate::parsing::FromStringOptional<$name> for $name {
+            fn from_str_optional(s : &str) -> $crate::error::Result<$name> {
+                s.parse::<$name>()
+            }
+        }
+
+        impl ::std::str::FromStr for $name {
+            type Err = $crate::error::Error;
+
+            fn from_str(s: &str) -> $crate::error::Result<$name> {
+                match s {
+                    $(
+                        $value => Ok($name::$variant),
+                    )*
+                    _ => Err($crate::error::Error::with_message($crate::error::ErrorKind::DataConversion, || format!("unknown variant of {} found: \"{}\"",
+                        stringify!($name),
+                         s
+                    )))
+                }
+            }
+        }
+
+        impl<'de> serde::Deserialize<'de> for $name {
+            fn deserialize<D>(deserializer: D) -> ::core::result::Result<Self, D::Error>
+            where
+                D: serde::Deserializer<'de>,
+            {
+                let s = String::deserialize(deserializer)?;
+
+                match s.as_ref() {
+                    $(
+                        $value => Ok(Self::$variant),
+                    )*
+                    _ => Err(serde::de::Error::custom("unsupported value")),
+                }
+            }
+        }
+
+        impl serde::Serialize for $name {
+            fn serialize<S>(&self, s: S) -> ::core::result::Result<S::Ok, S::Error>
+            where S: serde::Serializer {
+                return s.serialize_str(&self.to_string())
+            }
+        }
```

Comment:
nit: I think we could do this without implementing `Deserialize` and `Serialize` directly. My understanding is that applying [the `#[serde(rename)]` attribute](https://serde.rs/variant-attrs.html#rename) to each variant would do the same thing.

Again, I know you're just moving code so this is non-blocking.

Reply:
Yeah, I've never been sure why the author did this originally. Maybe to product faster code than `serde/derive`? I can't imagine it being much different. In my derive macro plans, I do plan to utilize `#[serde(rename)]`, though.